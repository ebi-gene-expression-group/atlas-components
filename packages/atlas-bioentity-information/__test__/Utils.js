const mockBioentityProperties = () => {
  return [ {
    type: `go`,
    name: `Gene Ontology`,
    values: [
      {
        text: `Link of relevance 1`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 1
      },
      {
        text: `a Link of relevance 2`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 2
      },
      {
        text: `Link of relevance 3`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 3
      },
      {
        text: `b Link of relevance 2`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 2
      },
      {
        text: `Link of relevance 3- another`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 3
      },
      {
        text: `RNA polymerase II core promoter proximal region sequence-specific DNA binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 0
      }, {
        text: `nucleic acid binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0003676`,
        relevance: 0
      }, {
        text: `intracellular`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0005622`,
        relevance: 0
      }, {
        text: `nucleus`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0005634`,
        relevance: 0
      }, {
        text: `transcription, DNA-templated`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0006351`,
        relevance: 0
      }, {
        text: `regulation of transcription, DNA-templated`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0006355`,
        relevance: 0
      }, {
        text: `zinc ion binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0008270`,
        relevance: 0
      }, {
        text: `metal ion binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0046872`,
        relevance: 0
      }]
    },
    {
      type: `synonym`,
      name: `Synonym`,
      values: [
        {
          text: `Sftp1`,
          url: ``,
          relevance: 0
        },
        {
          text: `Sftp-1`,
          url: ``,
          relevance: 0
        },
        {
          text: `SP-A`,
          url: ``,
          relevance: 0
        },
        {
          text: `surfactant pulmonary associated protein A1`,
          url: ``,
          relevance: 0
        },
        {
          text: `SFTPA1`,
          url: ``,
          relevance: 0
        }
      ]
    },
    {
      type: `ensgene`,
      name: `Ensembl gene`,
      values: [
        {
          text: `ENSMUSG00000021789`,
          url: `http://www.ensemblgenomes.org/id-gene/ENSMUSG00000021789`,
          relevance: 0
        }
      ]
    },
    {
      type: `enstranscript`,
      name: `Ensembl transcript`,
      values: [
        {
          text: `ENSMUST00000170719`,
          url: `http://www.ensemblgenomes.org/id/ENSMUST00000170719`,
          relevance: 0
        },
        {
          text: `ENSMUST00000022314`,
          url: `http://www.ensemblgenomes.org/id/ENSMUST00000022314`,
          relevance: 0
        }
      ]
    }
  ]
}

export { mockBioentityProperties }