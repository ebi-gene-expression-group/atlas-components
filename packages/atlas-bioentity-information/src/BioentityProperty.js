import React from 'react'
import PropTypes from 'prop-types'

class BioentityProperty extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            showAll: false
        }
    }

    // take three most relevant links and then all of the same relevance
    _pickMostRelevant(properties) {
        const relevanceThreshold =
            properties
                .map(p => p.relevance)
                .sort((l,r) => r-l)
                .concat([0,0,0])
                [properties.size < 3 ? properties.size-1 : 2];
        return properties.filter(p => p.relevance>=relevanceThreshold);
    }

    _renderProperty(property, ix) {
        return (
            property.url
                ? <a key={property.url+" "+ix} className={"bioEntityCardLink"} href={property.url} target="_blank">
                    {property.text}
                </a>
                : <span key={property.text + " "+ix}>
          {property.text}
        </span>
        )
    }

    _zipWithCommaSpans(elts) {
        return (
            [].concat.apply([],
                elts.map((e, ix)=>[e, <span key={"comma "+ix}>, </span>])
            )
                .slice(0,-1)
        )
    }

    render() {
        const numUnshownLinks = this.props.values.length - this._pickMostRelevant(this.props.values).length;
        const hasOptionalLinks = ["go","po"].indexOf(this.props.type) > -1 && numUnshownLinks >0;
        const showMoreLessButton = <a role="button" style={{cursor:"pointer"}} onClick={function(){
                                        this.setState((previousState)=>({showAll: !previousState.showAll}))
                                      }.bind(this)}>
                                      {this.state.showAll ? " (show less)" : " â€¦ and "+numUnshownLinks+" more"}
                                  </a>;

        return (
          <tr>
            <td className={"gxaBioentityInformationCardPropertyType"}>
                {this.props.name}
            </td>
            <td>
            <div>
              {hasOptionalLinks
                ?
                <span>
                  {
                    this._zipWithCommaSpans(
                      (this.state.showAll
                        ? this.props.values
                        : this._pickMostRelevant(this.props.values)
                      )
                        .sort((l,r)=>(
                          r.relevance === l.relevance
                            ? r.text.toLowerCase() < l.text.toLowerCase() ? 1 : -1
                            : r.relevance - l.relevance
                          )
                        ).map(this._renderProperty)
                    )}

                  {showMoreLessButton}
                </span>
                : this._zipWithCommaSpans(this.props.values.map(this._renderProperty))
              }

                </div>
            </td>
          </tr>
        )
    }
}

BioentityProperty.propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    values: PropTypes.arrayOf(PropTypes.shape({
        text: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
        relevance: PropTypes.number.isRequired
    }))
}

export default BioentityProperty