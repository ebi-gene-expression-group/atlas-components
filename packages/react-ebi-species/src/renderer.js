"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
