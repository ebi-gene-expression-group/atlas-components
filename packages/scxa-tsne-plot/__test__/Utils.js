import HighchartsSeriesGenerator from 'highcharts-series-generator'

const seriesNames = [`0`, `1`, `2`, `3`, `4`]
const maxPointsPerSeries = 1000

const gradientColourRanges = () => {
  return [
    {
      colour: `rgb(215, 255, 255)`,
      threshold: 0,
      stopPosition: 0
    },
    {
      colour: `rgb(128, 255, 255)`,
      threshold: 10,
      stopPosition: 20
    },
    {
      colour: `rgb(0, 85, 225)`,
      threshold: 100,
      stopPosition: 40
    },
    {
      colour: `rgb(0, 0, 115)`,
      threshold: 10000,
      stopPosition: 100
    }
  ]
}

const plotData = (series) => {
  const allPoints = series.reduce((acc, series) => acc.concat(series.data), [])

  return {
    series: series,
    min: Math.min(...allPoints.map((point) => point.expressionLevel)),
    max: Math.max(...allPoints.map((point) => point.expressionLevel)),
    unit: `TPM`
  }
}

const randomHighchartsSeries = () => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithNamesAndMaxPoints = (seriesNames, maxPointsPerSeries) => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithSeed = () => {
  const seed = `A hair, Morty. I need one of your hairs. This isnâ€™t Game of Thrones.`

  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries, seed)
}

export { plotData, gradientColourRanges, randomHighchartsSeries, randomHighchartsSeriesWithSeed, randomHighchartsSeriesWithNamesAndMaxPoints}