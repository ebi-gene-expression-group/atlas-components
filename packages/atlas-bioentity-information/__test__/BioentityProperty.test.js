import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import BioentityProperty from '../src/BioentityProperty'
import {mockBioentityProperties} from './Utils'

Enzyme.configure({ adapter: new Adapter() })

describe(`BioentityProperty`, () => {
  test(`with data matches snapshot`, () => {
    const bioentityProperty = mockBioentityProperties()[0]
    const tree = renderer.create(<BioentityProperty type={bioentityProperty.type} name={bioentityProperty.name} values={bioentityProperty.values}/>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test(`showMore button is not shown for other properties except gene ontology`, () => {
    const bioentityPropertyWithShowMore = mockBioentityProperties()[0]
    const bioentityPropertyWithoutShowMore = mockBioentityProperties()[1]

    const wrapperWithButton = mount(<BioentityProperty type={bioentityPropertyWithShowMore.type} name={bioentityPropertyWithShowMore.name} values={bioentityPropertyWithShowMore.values}/>)
    expect(wrapperWithButton.find(`a`).find({role: `button`}).exists()).toBe(true)

    const wrapperWithouButton = mount(<BioentityProperty type={bioentityPropertyWithoutShowMore.type} name={bioentityPropertyWithoutShowMore.name} values={bioentityPropertyWithoutShowMore.values}/>)
    expect(wrapperWithouButton.find(`a`).find({role: `button`}).exists()).toBe(false)
  })

  test(`the number of links increases when showMore is clicked`, () => {
    const bioentityProperty = mockBioentityProperties()[0]

    const wrapper = mount(<BioentityProperty type={bioentityProperty.type} name={bioentityProperty.name} values={bioentityProperty.values}/>)

    // Dummy data has more than 3 values so maximum 3 should always be displayed
    expect(wrapper.find(`.bioEntityCardLink`)).toHaveLength(3)

    // Simulate button click
    wrapper.find(`a`).find({role: `button`}).simulate('click')

    expect(wrapper.find(`.bioEntityCardLink`)).toHaveLength(bioentityProperty.values.length)
  })

  test(`properties without URLs are displayed as text`, () => {
    const bioentityProperty = mockBioentityProperties()[1]

    const wrapper = mount(<BioentityProperty type={bioentityProperty.type} name={bioentityProperty.name} values={bioentityProperty.values}/>)

    expect(wrapper.find(`.bioentityCardLink`).exists()).toBe(false)
  })

})