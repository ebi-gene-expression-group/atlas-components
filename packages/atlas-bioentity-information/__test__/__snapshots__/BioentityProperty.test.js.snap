// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BioentityProperty with data matches snapshot 1`] = `
<div>
  <span>
    <a
      className="bioEntityCardLink"
      href="http://amigo.geneontology.org/amigo/term/GO%3A0000978"
      target="_blank"
    >
      Link of relevance 1
    </a>
    , 
  </span>
  <span>
    <a
      className="bioEntityCardLink"
      href="http://amigo.geneontology.org/amigo/term/GO%3A0000978"
      target="_blank"
    >
      a Link of relevance 2
    </a>
    , 
  </span>
  <span>
    <a
      className="bioEntityCardLink"
      href="http://amigo.geneontology.org/amigo/term/GO%3A0000978"
      target="_blank"
    >
      Link of relevance 3
    </a>
    
  </span>
  <a
    onClick={[Function]}
    role="button"
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
     â€¦ and 10 more
  </a>
</div>
`;
